lib/Driver/ToolChains/Hcc.cpp

```diff
@@ -137,74 +153,155 @@ void HCC::CXXAMPAssemble::ConstructJob(Compilation &C, const JobAction &JA,
+    }
 }
 
-void HCC::CXXAMPLink::ConstructJob(Compilation &C,
-                                   const JobAction &JA,
-                                   const InputInfo &Output,
-                                   const InputInfoList &Inputs,
-                                   const ArgList &Args,
-                                   const char *LinkingOutput) const {
+void HCC::CXXAMPLink::ConstructJob(
+    Compilation &C,
+    const JobAction &JA,
+    const InputInfo &Output,
+    const InputInfoList &Inputs,
+    const ArgList &Args,
+    const char *LinkingOutput) const
+{
   ArgStringList CmdArgs;
 
   // add verbose flag to linker script if clang++ is invoked with --verbose flag
   if (Args.hasArg(options::OPT_v))
     CmdArgs.push_back("--verbose");
 
   // specify AMDGPU target
+
   if (Args.hasArg(options::OPT_amdgpu_target_EQ)) {
-    auto AMDGPUTargetVector = Args.getAllArgValues(options::OPT_amdgpu_target_EQ);
+    auto AMDGPUTargetVector =
+        Args.getAllArgValues(options::OPT_amdgpu_target_EQ);
 
     for (auto&& AMDGPUTarget : AMDGPUTargetVector) {
       // TODO: the known GFXip list should probably reside in a constant
       //       global variable so as to allow easy extension in the future.
-      static constexpr const char prefix[] = "--amdgpu-target=";
-      static constexpr unsigned int discard_sz = 3u;
-      static const std::string gfx_ip{"gfx"};
       static const std::string long_gfx_ip_prefix{"AMD:AMDGPU:"}; // Temporary.
 
       // TODO: this is temporary.
-      if (std::search(AMDGPUTarget.cbegin(), AMDGPUTarget.cend(), long_gfx_ip_prefix.cbegin(), long_gfx_ip_prefix.cend()) != AMDGPUTarget.cend()) {
-          AMDGPUTarget = temporaryReplaceLongFormGFXIp(C, AMDGPUTarget);
+      if (std::search(
+              AMDGPUTarget.cbegin(),
+              AMDGPUTarget.cend(),
+              long_gfx_ip_prefix.cbegin(),
+              long_gfx_ip_prefix.cend()) != AMDGPUTarget.cend()) {
+          AMDGPUTarget = temporary_replace_long_form_GFXIp(C, AMDGPUTarget);
       }
 
-      if (std::search(AMDGPUTarget.cbegin(), AMDGPUTarget.cend(), gfx_ip.cbegin(), gfx_ip.cend()) != AMDGPUTarget.cend()) {
-        std::string t{prefix};
-        switch (std::atoi(AMDGPUTarget.data() + discard_sz)) {
-        case 700: t += "gfx700";  break;
-        case 701: t += "gfx701";  break;
-        case 801: t += "gfx801"; break;
-        case 802: t += "gfx802";   break;
-        case 803: t += "gfx803";    break;
-        case 900: t += "gfx900";  break;
-        case 901: t += "gfx901";  break;
-        default:
-          C.getDriver().Diag(diag::warn_amdgpu_target_invalid) << AMDGPUTarget;
-        break;
-        }
-        CmdArgs.push_back(Args.MakeArgString(t));
+      validate_and_add_to_command(AMDGPUTarget, C, Args, CmdArgs);
+    }
+  }
+  else {
+      constexpr const char null_agent[] = "gfx000";
+      const auto detected_targets = detect_gfxip(C, getToolChain());
```

Comment:
@AlexVlx perhaps it's worthwhile to add some comments here to shed some lights on the expected behavior of `rocm_agent_enumerator`?